### ISSUE TITLE
halide_handle_traits::type_info() needs rethinking

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#1407]
It relies on instantiation of static template members having identical pointers in situations that we can't rely on that (this already bit us for IRNode); it appears to be working at the present time but it's likely that we're just getting lucky. Opening this issue so that we don't wait until random clang/llvm variations make this bite us.

**Comment by zvookin** [comment]
@steven-johnson @abadams Didn't we come up with a fix for this?

**Comment by abadams** [comment]
No, we changed halide_type_of, but the traits structs for pointer types are
the same.
